[jit] Add RRef to IValue and Jit

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30137 [jit] Add RRef to IValue and Jit**
* #30630 [rpc] Remove template on RRef and add Type to RRef creation
* #30136 [jit] Enable IValue to hold a PyObject

Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233)

wanchaol committed:
[WIP] Add RRef to IValue and Jit

[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit d0c52de:
* **2/2** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=d0c52de05d31325a04ad586e487feed71d032cb0).

### 2 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3960685) pytorch_linux_xenial_py3_6_gcc5_4_build (1/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74869035))
<details>
<summary>
<code>Automatic merge failed; fix conflicts and then commit the result.</code>
</summary>

```
  git checkout -b <new-branch-name> 
 
HEAD is now at d0c52de05d Update on "[WIP] Add RRef to IValue and Jit" 
+ git reset --hard d0c52de05d31325a04ad586e487feed71d032cb0 
HEAD is now at d0c52de05d Update on "[WIP] Add RRef to IValue and Jit" 
+ git merge --allow-unrelated-histories --no-edit --no-ff 409151e1bb90c66546901d372f59930f72242ee9 
Auto-merging torch/csrc/jit/script/script_type_parser.cpp 
CONFLICT (content): Merge conflict in torch/csrc/jit/script/script_type_parser.cpp 
Auto-merging torch/csrc/distributed/rpc/py_rref.cpp 
Auto-merging tools/build_variables.py 
Automatic merge failed; fix conflicts and then commit the result. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3960677) pytorch_linux_xenial_py2_7_9_build (2/2)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74869050))
<details>
<summary>
<code>Dec 16 08:16:38 /var/lib/jenkins/workspace/aten/src/ATen/core/ivalue.cpp:30:9: error: enumeration value 'RRef' not handled in switch [-Werror=switch]</code>
</summary>

```
Dec 16 08:16:25 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/function_schema.cpp.o 
Dec 16 08:16:28 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/LegacyDeviceTypeInit.cpp.o 
Dec 16 08:16:29 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/type.cpp.o 
Dec 16 08:16:31 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/interned_strings.cpp.o 
Dec 16 08:16:31 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/Dimname.cpp.o 
Dec 16 08:16:31 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/LegacyTypeDispatch.cpp.o 
Dec 16 08:16:31 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/ivalue.cpp.o 
Dec 16 08:16:32 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/VariableHooksInterface.cpp.o 
Dec 16 08:16:38 [ 49%] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/dispatch/Dispatcher.cpp.o 
Dec 16 08:16:38 /var/lib/jenkins/workspace/aten/src/ATen/core/ivalue.cpp: In member function 'c10::TypePtr c10::IValue::type() const': 
Dec 16 08:16:38 /var/lib/jenkins/workspace/aten/src/ATen/core/ivalue.cpp:30:9: error: enumeration value 'RRef' not handled in switch [-Werror=switch] 
Dec 16 08:16:38    switch(tag) { 
Dec 16 08:16:38          ^ 
Dec 16 08:16:38 cc1plus: all warnings being treated as errors 
Dec 16 08:16:38 caffe2/CMakeFiles/torch_cpu.dir/build.make:1066: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/ivalue.cpp.o' failed 
Dec 16 08:16:38 make[2]: *** [caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/core/ivalue.cpp.o] Error 1 
Dec 16 08:16:38 make[2]: *** Waiting for unfinished jobs.... 
Dec 16 08:16:44 make[1]: *** [caffe2/CMakeFiles/torch_cpu.dir/all] Error 2 
Dec 16 08:16:44 CMakeFiles/Makefile2:5332: recipe for target 'caffe2/CMakeFiles/torch_cpu.dir/all' failed 
Dec 16 08:16:44 make: *** [all] Error 2 
Dec 16 08:16:44 Makefile:138: recipe for target 'all' failed 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

pritamdamania87 commented on aten/src/ATen/core/RRef.h:
```diff
@@ -0,0 +1,42 @@
+
+// This abstract class contains only user-facing APIs, and will be shared
+// between jit and distributed to implement TorchScript support.
+class C10_EXPORT RRef : public c10::intrusive_ptr_target {
```
Shouldn't this class be called `RRefInterface` and we can call `RRefBase` simply `RRef`? Also, should the comment block in RRef.h explaining RRefs move here?

wanchaol replied:
yeah I was just pulling the changes from https://github.com/pytorch/pytorch/pull/28942 to test it initially. `RRefInterface` is probably a better name I agree.

pritamdamania87 commented on aten/src/ATen/core/RRef.h:
```diff
@@ -0,0 +1,42 @@
+  // Returns true if this is the ``OwnerRRef``
+  virtual bool isOwner() const = 0;
+
+  virtual TypePtr type() const = 0;
```
nit: comment here, also TypeKind vs TypePtr?

pritamdamania87 commented on torch/csrc/distributed/rpc/py_rref.h:
```diff
@@ -25,7 +24,7 @@ class PyRRef {
 
  private:
-  std::shared_ptr<RRef> rref_;
+  std::shared_ptr<RRefBase> rref_;
```
Shouldn't this be the abstract class type?

pritamdamania87 commented on torch/csrc/distributed/rpc/rref_impl.h:
```diff
@@ -184,15 +185,15 @@ struct RRefForkData {
 // ``RRef`` is the base type for both ``UserRRef`` and ``OwnerRRef``.
 // Each ``RRef`` has a globally unique ``RRefId``.
-class RRef {
+class RRefBase : public c10::RRef {
```
nit: add `override` for all overridden methods in this class.

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[WIP] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
Update on "[jit] Add RRef to IValue and Jit"




[ghstack-poisoned]

wanchaol committed:
fix clang-tidy on "[jit] Add RRef to IValue and Jit"


Differential Revision: [D19502233](https://our.internmc.facebook.com/intern/diff/D19502233)

[ghstack-poisoned]

wanchaol commented:
outdated PR and bit hard to rebase this, close this in favor of https://github.com/pytorch/pytorch/pull/32992
